### ISSUE TITLE
User passwords handle by admin

### ISSUE THREAD

**Issue by manfer** [gogits/gogs#1707]
I don't like at all how admins can handle passwords in admin panel.

* They should be able to create users and assign a password to them. This is already available but... in the add user they should have a password confirmation field to avoid misstype errors.

* Only the owner of the account should be able to change password. The owner of the gogs installation could still change it directly on the database at a last resort (better never do that) but should not be possible to do it on the admin panel. Not even by admins (unless they know the password). This means that the edit user form should be changed. Instead of just a password field it should include three fields, old password field, new password and confirmation password field to handle password change. There is already a forgot password feature and at a last resort the password can be overwritten on database (should not be done anyway unless no other option, account lost could make more sense in some situations). There is no need to allow admins to change passwords. A service were you can ask an admin to reset the password looks unprofessional to me and a target for social engineering by the way.

* Forgot password only works if confirmation is enable. Not sure why this limitation (have not tested if indeed the message is accurate an that's what is happening). If someone has used a fake email and the admin has decided to allow any email (no confirmation needed) and then that user forgets his password it is that user problem. The mail server is clever enough to try to deliver mail. It should not be a gogs task to try to handle mail tasks. Should be totally agnostic about mail handling. If this is some way to try to avoid any type of abuse it is not gogs task either in my opinion is a task for the mail server and related security services on the server. Users that use valid emails even if gogs owner has opt for email confirmation off should be able to try to reset their passwords with forget password feature. Different is if mailer is not configured. In that case gogs can't send mail and can't send a reset password mail but then the reason is not confirmation not enabled but mailer not avialable.

* If mailer is not configured gogs should not show a forget password link because it is not possible to send a reset password mail. Instead of having the link and on click showing a message (wrong message by the way) makes much more sense to totally remove the forget password link in this situation.

**Comment by Unknwon** [comment]
Thanks your feedback!

> They should be able to create users and assign a password to them. This is already available but... in the add user they should have a password confirmation field to avoid misstype errors.

If misstyped, just reset again.

> Only the owner of the account should be able to change password. The owner of the gogs installation could still change it directly on the database at a last resort (better never do that) but should not be possible to do it on the admin panel. Not even by admins (unless they know the password). This means that the edit user form should be changed. Instead of just a password field it should include three fields, old password field, new password and confirmation password field to handle password change. There is already a forgot password feature and at a last resort the password can be overwritten on database (should not be done anyway unless no other option, account lost could make more sense in some situations). There is no need to allow admins to change passwords. A service were you can ask an admin to reset the password looks unprofessional to me and a target for social engineering by the way.

So how do you reset user password when e-mail service is not enabled(can't reset by e-mail)?

> If mailer is not configured gogs should not show a forget password link because it is not possible to send a reset password mail. Instead of having the link and on click showing a message (wrong message by the way) makes much more sense to totally remove the forget password link in this situation.

Agree.

**Comment by manfer** [comment]
> So how do you reset user password when e-mail service is not enabled(can't reset by e-mail)?

This is the question. You don't. It is a user resposability to keep his password. Reset password is a user resposibility too and not an admin resposability (amount other things because otherwise admins are a target for social engineering).

Reset by email, security questions, mobile SMS, that gives access to a one time login or one time reset form. Don't know if I miss any but those are the methos used nowadays.

As I bet mobile SMS is not possible that would mean gogs would need security questions as an alternative reset password method.

**Comment by Unknwon** [comment]
> This is the question. You don't. It is a user resposability to keep his password. Reset password is a user resposibility too and not an admin resposability (because otherwise admins are a target for social engineering).

Well, a question about how to make everyone happy again, Gogs used to not provide reset password by admin by someone asked.

> Reset by email, security questions, mobile SMS, that gives access to a one time login or one time reset form. Don't know if I miss any but those are the methos used nowadays.

Good point, security questions can be another option.

**Comment by manfer** [comment]
> Well, a question about how to make everyone happy again, Gogs used to not provide reset password by admin by someone asked.

Sorry :) Well, so maybe just an ADMINS_CAN_RESET_PASSWORDS on/off admin flag.

**Comment by Fastidious** [comment]
I disagree. If an admin falls for social engineering and gets its password compromised, it shouldn't run Gogs. I have root in all my VPS, and have never gotten it compromised. 

A flag, I guess, could make *everyone* happy.

**Comment by Unknwon** [comment]
Thank you both @manfer @Fastidious !

So things to do:

- Add config option such as `ADMINS_CAN_RESET_PASSWORDS`
- Able to set security questions to reset password

Any comments?

**Comment by manfer** [comment]
> I disagree. If an admin falls for social engineering and gets its password compromised

It is not his password what would be compromised. It is other users passwords.

If I email an admin and I pretend to be Marty McFly and ask him a password reset because I lost access to my registration email and I give him an email to send the new password and he sends it to me that wouldn't do any good. That's the social engineering I'm talking about.

If he is the admin and he leaks his password. Well, there isn't a nice word to describe that. LOL

@Unknwon That sounds good
